test/hotspot/jtreg/runtime/Monitor/ConcurrentDeflation.java

```diff
@@ -0,0 +1,83 @@
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+import jdk.test.lib.process.OutputAnalyzer;
+import jdk.test.lib.Platform;
+import jdk.test.lib.process.ProcessTools;
+
+import java.lang.management.ManagementFactory;
+import java.lang.management.ThreadInfo;
+import java.lang.management.ThreadMXBean;
+
+/*
+ * @test
+ * @bug 8318757
+ * @summary Test concurrent monitor deflation by MonitorDeflationThread and thread dumping
+ * @library /test/lib
+ * @run main/othervm -XX:+UnlockDiagnosticVMOptions -XX:GuaranteedAsyncDeflationInterval=2000 -XX:LockingMode=0 ConcurrentDeflation
+ */
+
+public class ConcurrentDeflation {
+    public static final int TOTAL_RUN_TIME = 10 * 1000;
+
+    public static void main(String[] args) throws Exception {
+        Thread thread_dumper  = new Thread(() -> dumpThreads());
+        thread_dumper.start();
+        Thread monitor_creator  = new Thread(() -> createMonitors());
+        monitor_creator.start();
+
+        thread_dumper.join();
+        monitor_creator.join();
+    }
+
+    static private void dumpThreads() {
+        ThreadMXBean threadBean = ManagementFactory.getThreadMXBean();
+        long startTime = System.currentTimeMillis();
+        int dumpCount = 0;
+
+        while (System.currentTimeMillis() - startTime < TOTAL_RUN_TIME) {
+            threadBean.dumpAllThreads(true, false);
+            dumpCount++;
+            try {
+                Thread.sleep(10);
+            } catch (InterruptedException e) {}
+        }
+        System.out.println("Dumped all thread info " + dumpCount + " times");
+    }
+
+    static private void createMonitors() {
+        Object[] monitors = new Object[1000];
+        int monitorCount = 0;
+
+        int index = 0;
+        long startTime = System.currentTimeMillis();
+        while (System.currentTimeMillis() - startTime < TOTAL_RUN_TIME) {
+            index = index++ % 1000;
+            monitors[index] = new Object();
+            synchronized (monitors[index]) {
+            }
```

Comment:
I would expect C2 to eliminate this as well. The monitors are provably thread-local so synchronization is a no-op.

Reply:
Done

Reply:
It still might eliminate locking due to the empty block.

Reply:
Yes, I agree with David. Look how other monitor inflation tests do this, we call `wait()` to guarantee both inflation (although it is partially handled by `LockingMode=0` here) and escaping to native call thus avoiding aggressive compiler opts.
 https://github.com/openjdk/jdk/blob/6b21ff61dad6f633c744c1c33c29ea86183b509d/test/hotspot/jtreg/runtime/Monitor/DeflationIntervalsTest.java#L132-L140

Reply:
Right, but currently it doesn't remove the synchronized block in my runs and this test was mainly added to be a very specific test that provokes the bug that this PR fixes. I previously considered adding a wait(1), but that will significantly limit the number of created monitors. Maybe I can add a counter inside the synchronized block, just like our JMH lock micros do?